_posts/2021-06-07-spring-schedulers.md

```diff
@@ -0,0 +1,281 @@
+    long delay);
+
+}
+```
+
+**Now every task is run as a thread on the thread pool and spring internally manages it as a scheduled task which Internally has a task and which is runnable at its core**
+
+```java
+//this is protected, so you cannot  create an instance of this manually 
+public final class ScheduledTask {
+ private final Task task;
+ @Nullable
+ volatile ScheduledFuture<?> future;
+
+ ScheduledTask(Task task) {
+   this.task = task;
+ }
+ public void cancel() {
+   ScheduledFuture<?> future = this.future;
+   if (future != null) {
+     future.cancel(true);
+   }
+ }
+}
+
+public class Task {
+ private final Runnable runnable;
+
+ public Task(Runnable runnable) {
+   Assert.notNull(runnable, "Runnable must not be null");
+   this.runnable = runnable;
+ }
+
+ public Runnable getRunnable() {
+   return this.runnable;
+ }
+}
+```
+
+Now hopefully all basics are clear, lets where all the magic happens and how these things come together to save the world
+
+```java
+ScheduledAnnotationBeanPostProcessor
+```
+ is the class, when you use `@Scheduled` annotation on any void method, this class holds the responsibility of registering that as ScheduledTask and its management.
+
+This class has a `processScheduled()` which is called when `@Scheduled` is used,
+And this class register the task, based on whatever type of flavour you have used for example **if its a `@Scheduled(cron=””)`  it will register a CronTask, if it is a `@Scheduled(fixedRate=””)` it will register a FixedRateTask.**
+
+```java
+public class CronTask extends TriggerTask {
+ private final String expression;
+
+ public CronTask(Runnable runnable, String expression) {
+   this(runnable, new CronTrigger(expression));
+ }
+
+ public CronTask(Runnable runnable, CronTrigger cronTrigger) {
+   super(runnable, cronTrigger);
+   this.expression = cronTrigger.getExpression();
+ }
+
+ public String getExpression() {
+   return this.expression;
+ }
+}
+``` 
+
+Now we won’t look at all the tasks, you can look at the processScheduled method for more detail.
+Why I wanted to show is to have your attention on **private final String expression;** 
+
+**It is final which means once we use `@Scheduled(cron=””)`, we CANNOT update it programmatically later**, if we want to change it, we will need to register a new task and stop this one.
+
+The Same goes with tasks of any other flavor, **they all are IMMUTABLE when we create it with `@Scheduled` annotation**.
+
+**Unfortunately, no annotation ssupports mutable tasks**. That means we will have to do it manually.
+
+Now to have the ability to change the scheduling at run time, we will need to use  
+```java
+ScheduledFuture schedule(Runnable task, Trigger trigger);
+```             
+Recollect we already saw this in `TaskScheduler` and that's what we will see in the next section.
+
+
+## How to schedule a task without annotation
+
+  The first question that comes to any developers mind is **Why schedule a task with annotation?**        
```

Comment:
Is that supposed to mean "without annotation"?

Reply:
yes corrected